### DISCUSSION TITLE
The ai assistant of supabase sql editor is not working in the local development.

### DISCUSSION THREAD

**Discussion by devhkr** [supabase/supabase#21621] [Questions]
When using AI ASSISTANT in a local environment, the following error occurs
<img width="1374" alt="스크린샷 2024-02-27 오후 10 33 18" src="https://github.com/supabase/supabase/assets/6085416/287b4d96-9722-47ac-99c4-f7c1748634a7">

Looking at the logs, I see the following error
<img width="1882" alt="스크린샷 2024-02-27 오후 10 33 49" src="https://github.com/supabase/supabase/assets/6085416/a6d4469d-c2d6-4b2b-b8e5-0b58187700c7">

`2024-02-27 22:32:52 OpenAIError: The OPENAI_API_KEY environment variable is missing or empty; either provide it, or instantiate the OpenAI client with an apiKey option, like new OpenAI({ apiKey: 'My API Key' }).
2024-02-27 22:32:52     at new OpenAI (file:///app/node_modules/openai/index.mjs:26:19)
2024-02-27 22:32:52     at /app/apps/studio/.next/server/pages/api/ai/sql/edit.js:1:1710`

Doesn't AI ASSISTANT work in the local environment by default? Or is there something wrong with my settings?
If the settings are wrong, I'm wondering how I can fix them.  :)

**Comment by charislam** [comment]
Hey @devhkr! The clue is in this part of the error message: `The OPENAI_API_KEY environment variable is missing or empty`

You'll need to provide your own `OPENAI_KEY` in your env file.
(Note Open AI is expecting the default `OPENAI_API_KEY`, but in our code we've renamed it `OPENAI_KEY`, so that's what you'll have to name yours as well.)

**Reply by devhkr** [reply]
HI @charislam  Thanks for your response.
Can you explain which environment file you put OPENAI_API_KEY in? Or can you give me the URL to the documentation that explains this?

**Reply by charislam** [reply]
As part of the setup you should have copied the environment file from `.env` to `.env.local`, you can add the key there.

**Reply by charislam** [reply]
@devhkr, I've been deleting your follow-up comments because I'm not sure if you intended for that screenshot to expose as much of your key as it did, so I decided to err on the side of safety. Ignore me and consider me overly paranoid if that was just a dummy key 😄 

Original followup without screenshot:
> Sorry for the constant questions. 😢
> I've modified the one parameter in the .env.local file as follows, and did the same for the .env file just in case.
> Then we restarted supabase.
> However, I still get the same error.
> Did I do something wrong?

**Reply by devhkr** [reply]
I've uploaded the screenshots for clarity for those of you who are aware of this situation.
I removed the sensitive information. :) 

I followed the following documentation to set up my development environment.
https://supabase.com/docs/guides/cli/local-development

And as shown above, I wrote both OPENAI_API_KEY and OPENAI_KEY in the env.local file.
I also wrote them to the .env file just in case.
Next, I restarted supabase and got the same error.

Am I doing something wrong?

**Reply by charislam** [reply]
Hey @devhkr! Thanks for your patience!

Apparently I misunderstood your original question: I thought you were doing local development on studio, but if I understand correctly, you're running Supabase locally with the CLI to do local development of your own app. 🤦🏼‍♀️ Sorry, I should've looked at your original screenshots a bit more carefully.

In that case, I believe the solution is to just pass the environment variable when running `supabase start` (see this [PR that added the feature](https://github.com/supabase/cli/pull/1986)). Let me know if that works for you!

**Reply by devhkr** [reply]
Hi @charislam 
I apologize for any confusion I may have caused by not explaining the situation in enough detail. 😔
I installed the CLI with homebrew and the latest version is 1.145.4.
That PR seems to have been applied in v1.147.3.
I think that's why the env file didn't work.
So it looks like the supabase team will need to update the supabase cli package for homebrew.
<img width="607" alt="스크린샷 2024-03-05 오후 6 07 36" src="https://github.com/supabase/supabase/assets/6085416/5826c455-faa0-414a-aadc-b89a02b2c50c">

**Reply by devhkr** [reply]
Hi @charislam 
I upgraded the supabase package in homebrew and it works fine with no errors. 
`brew upgrade supabase`
Thanks for all your help. :)

**Reply by charislam** [reply]
Awesome, glad to hear it worked!

**Comment by mrcaidev** [comment]
In a nutshell, put `OPENAI_API_KEY=...` in `your-project/.env`.

A few catches that confused me for quite some time:

1. `.env.local` doesn't work in the time of writing. Only `.env` is supported.
2. Put `.env` in the **root** of your project, not in `supabase/`.

**Reply by HoekWax** [reply]
You did manage to make it work ? I did what you wrote, i then run `supabase start`.
But the AI assistant doesn't work locally, when I log the error using `docker logs supabase_studio_lvlup-supabase`, I get this error
```
⨯ OpenAIError: The OPENAI_API_KEY environment variable is missing or empty; either provide it, or instantiate the OpenAI client with an apiKey option, like new OpenAI({ apiKey: 'My API Key' }).
    at new OpenAI (file:///app/node_modules/openai/index.mjs:26:19)
    at /app/apps/studio/.next/server/pages/api/ai/sql/edit.js:1:1872
```

When i execute `docker exec -it supabase_studio_lvlup-supabase env` to check the env, I see that OPENAI_KEY exists in the env, it has the value from the key OPENAI_API_KEY that I set in my .env

It seems like the studio wants to use OPENAI_API_KEY but the value set in the env of the container is OPENAI_KEY, event if in my .env at the root folder I only set the key OPENAI_API_KEY and not the key OPENAI_KEY

Does anyone have the same issue ? Or know what's going on ?

**Comment by mkungla** [comment]
### Inconsistency and Issues with Environment Configuration

I am encountering several persistent issues that significantly impact the development workflow, specifically around environment configuration and error handling. I would appreciate any help or clarification on these matters:

#### 1. Inconsistent Naming for Environment Variables
There's an ongoing inconsistency in how environment variables are named (`OPENAI_API_KEY` vs. `OPENAI_KEY`). This can lead to confusion and potential errors in configuration.

#### 2. Loading .env Files in a Monorepo
The logic for loading `.env` files in a monorepo setup is unclear, especially when dealing with nested directories. For instance:
- `.env` is not recognized when placed in the root (`.`) or directly in the service directory (`./services/datastore/supabase`).
- It is instead only recognized when placed in `./services/datastore/.env`.

#### 3. Local Environment Error
Despite the `.env` files loading correctly, attempts to execute requests result in an error. Here is the error message received:
```json
{
  "error": "There was an unknown error editing the SQL snippet. Please try again."
}
```
#### 4. Concern About Closed Related Issues

I've noticed that previous bug reports related to these issues were closed without a clear resolution. Can someone explain the rationale behind these decisions? It would be beneficial to reopen these discussions to address and resolve the concerns effectively.

**Reply by rngadam** [reply]
Facing the same issues. First, error messsage in frontend talks about OPENAI_KEY, but what makes it work seems to be OPENAI_API_KEY: 

<img width="395" alt="Screenshot 2024-04-25 at 10 41 45 AM" src="https://github.com/supabase/supabase/assets/538542/827c0f13-e32f-4cd5-8ddd-9e5d45225df4">

after setting OPENAI_API_KEY in the root .env, I get a different, cryptic error message:

<img width="378" alt="Screenshot 2024-04-25 at 10 51 39 AM" src="https://github.com/supabase/supabase/assets/538542/ee7de31b-8426-4688-a23f-bf710ff5d7f1">

found out after looking up the logs from supabase/studio container that it is actually caused by exceeding quota:

```
$ docker logs e0559b085f10
  ▲ Next.js 14.2.0
  - Local:        http://localhost:3000
  - Network:      http://0.0.0.0:3000

 ✓ Starting...
 ✓ Ready in 1868ms
AI SQL generation failed: 429 You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.
AI SQL generation failed: 429 You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.
```

**Reply by hgup** [reply]
> ```
> $ docker logs e0559b085f10
>   ▲ Next.js 14.2.0
>   - Local:        http://localhost:3000
>   - Network:      http://0.0.0.0:3000
> 
>  ✓ Starting...
>  ✓ Ready in 1868ms
> AI SQL generation failed: 429 You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.
> AI SQL generation failed: 429 You exceeded your current quota, please check your plan and billing details. For more information on this error, read the docs: https://platform.openai.com/docs/guides/error-codes/api-errors.
> ```

+1 It exceeds rate limit. I don't know what limit it is referring to since I never used my OPEN AI API keys before

**Reply by HoekWax** [reply]
It's because you didn't set a credit card in you open ai account, you have to buy like 10 dollars or more of credits so you can use the api

**Reply by hgup** [reply]
Didn't know that... silly me... thanks

**Comment by devhkr** [comment]
I put OPENAI_API_KEY in the .env file and it worked fine.
But after upgrading the supabase CLI library, the error occurs again.
So I'm reopening the discussion. 🥲

```
OpenAIError: The OPENAI_API_KEY environment variable is missing or empty; either provide it, or instantiate the OpenAI client with an apiKey option, like new OpenAI({ apiKey: 'My API Key' }).
```

**Reply by HoekWax** [reply]
Do you know what version of the cli you were using previously ? Maybe we should use this one until this is fixed

**Reply by devhkr** [reply]
I think it was v1.147.3.

This is the version with the  [PR](https://github.com/supabase/cli/pull/1986) applied.

I upgraded to homebrew, so I'm not sure which version it

**Reply by HoekWax** [reply]
Hello @charislam, can you check if everything is alright with that ? It seems like a naming issue for the OpenAi api key still exists, I've described it on [another post](https://github.com/orgs/supabase/discussions/21621#discussioncomment-9548488) in this discussion

**Reply by devhkr** [reply]
I upgraded the supabase cli to 1.169.8 using homebrew, and it doesn't cause any errors, but the UI has changed like a service in the cloud, but the AI assistant doesn't work properly.
I thought it would work over time, so I decided not to use it for now.

**Reply by HoekWax** [reply]
I had another [issue](https://github.com/supabase/supabase/issues/26245) and they fixed it in beta release for now

So i tried starting the studio using the beta cli release by running `npx supabase@beta start` and it works !

**Reply by devhkr** [reply]
It hasn't been applied to the homebrew database yet, so I'll have to try it when it is.

**Reply by adamlamaa** [reply]
I had this issue and upgraded from supabase 1.167.4 -> 1.172.2 which resolved the issue and the AI assistante now works in my studio.

I have a .env with the OPENAI_API_KEY key in the root of my repo.

**Comment by ifallah** [comment]
i add

OPENAI_API_KEY=yourkey
OPENAI_KEY=yourkey

in (.env) & (docker-compose.yml)

and copye .env to root folders supabase 

and its work now

**Comment by Rocke1001feller** [comment]
# 1、summary
### --  before 2024/04/25 or before v1.24.04, 
## add `OPENAI_KEY` to .env or docker-compose.yml, any setting is OK .. and most important execute`docker compose up --force-recreate`
### --  after 2024/04/25 or after v1.24.04(include),
## use `OPENAI_API_KEY` to .env or docker-compose.yml, any setting is OK ..and most important execute `docker compose up --force-recreate`

# 2、below are details

![image](https://github.com/supabase/supabase/assets/97436317/dcd719e9-8078-4981-852d-87e77f1fff79)
![image](https://github.com/supabase/supabase/assets/97436317/fd7b6a85-9002-47f9-ae30-84a38e2579fb)

![image](https://github.com/supabase/supabase/assets/97436317/f1db2d30-caab-4c64-aa38-cb2c0bbce769)

![image](https://github.com/supabase/supabase/assets/97436317/75f54240-cbe0-4213-945c-8b0cd9a851c9)

![image](https://github.com/supabase/supabase/assets/97436317/a30fff9b-04ca-439d-aad7-d6526b35033b)


![image](https://github.com/supabase/supabase/assets/97436317/331f4a02-7882-495f-ad23-567e4e8e6616)

**Comment by adao** [comment]
why do i need an openai_api_key at all? any option to turn this off?

**Reply by Monu18765** [reply]
Yup good question

**Reply by davidtranjs** [reply]
same question, love to disable it by default

**Reply by zda** [reply]
Yes, it's not a necessary feature, and the constant toast messages obscure actual content on the screen.

**Comment by nicetomytyuk** [comment]
What model should I choose in OpenAI platform to make it work at best?

**Reply by HenkDz** [reply]
I'm here to find out if we can choose models, it uses gpt-3.5 by default a nightmare

**Comment by BrandiATMuhkuh** [comment]
I have started a new project today (13. Dec. 2024). And I got the following message when opening the AI editor side panel.
```
OpenAI API key not set
Add your OPENAI_API_KEY to ./docker/.env to use the AI Assistant.
```
<img width="458" alt="image" src="https://github.com/user-attachments/assets/424ac8a9-94d7-4c99-bf9e-2e9b5b70e65e" />

## What works
Putting the `.env` into the root of the mono repo (not supabase folder)

## What does NOT work
Putting the `.env` into the `./docker/.env` or `./supabase/docker/.env`


## suggestion
The editor should change the text `./docker/.env` to `.env`. Or be more clear where to put it depending on mono repo, etc.

**Reply by nicetomytyuk** [reply]
Were you able to make the Assistant work? After I set the `OPENAI_API_KEY` when I try to input anything in the chat I get code 500 in response with: `{"error":"There was an error processing your request"}`

**Reply by BrandiATMuhkuh** [reply]
Yes for me it works just fine. 
All I did is restart (stop -> start) supabase via cli

**Reply by nicetomytyuk** [reply]
Oh I am testing it in self-hosted environment

**Reply by damassi** [reply]
~I too get a 500, when attempting to run supabase locally. It reads the key just fine.~ 

Edit: Started working suddenly! not sure why.

**Comment by Diego-Romero** [comment]
https://github.com/orgs/supabase/discussions/21621#discussioncomment-11553912 This should be way clearer in the docs, there's no indication of this :/

**Comment by timurcatakli** [comment]
I am getting
```
Project `proj_Q90NpLLbVeXO3Rnm6ww9QfDA` does not have access to model `gpt-4.1-2025-04-14`
``` 

I am not sure how this is possible. My API key has full access. I created new keys twice. My key is in my `.env` which is in my root.

Any ideas?
Thanks

**Comment by vidicunt** [comment]
This still doesn't work for me in a self hosted docker instance
